### DISCUSSION TITLE
Need help with Data Table internationalization.

### DISCUSSION THREAD

**Discussion by fitimbytyqi** [shadcn-ui/ui#2636] [Q&A] [answered]
Hello, I have a case where I need to translate table headers based on the language selected.

Currently I'm using this approach.

1) I'm converting my columns.tsx to a function instead.

<img width="649" alt="Screenshot 2024-01-31 at 17 02 22" src="https://github.com/shadcn-ui/ui/assets/61948517/39497c72-3884-42c3-b54c-7d8a1921fa8d">

2) After that, I pass the whole function as a prop to Data Table (because we cannot call functions in server components) like so:
ps. This part is where I'm not sure if I'm doing it right performance wise.

<img width="606" alt="Screenshot 2024-01-31 at 17 02 35" src="https://github.com/shadcn-ui/ui/assets/61948517/a66547bc-ac2e-4ce7-a8b1-5be1d70a3ea3">

3) I call it inside the data table component and pass the translations as prop like so:

<img width="485" alt="Screenshot 2024-01-31 at 17 02 55" src="https://github.com/shadcn-ui/ui/assets/61948517/47b93d91-eb81-48fb-864f-16da44f23f62">

I also recently found this new idea https://github.com/shadcn-ui/ui/issues/409#issuecomment-1763536344

Which one is a better approach ?

Thanks a lot for your time.

**Comment by TheDevick** [comment]
Hello!

So, I know this issue was opened almost a year ago, but I'm facing the same issue as you.

Then I wanted to ask, what happened? Did you find a solution?

Thanks!

**Reply by fitimbytyqi** [reply]
Hey, I continued to use the approach I've highlighted above, but the way with metadata (the issue added above) should work too.

**Comment by sidimed-bahan** [comment]
Hi,

I know this issue was opened almost a year ago, and I am facing the same problem. However, the solution provided did not work for me, so I came up with a simpler alternative using next-intl. Here’s what I did:

1. In the en.json file (your messages file), I defined the header like this: 
{"clientName": "Client Name"}
2 . In the column.tsx file, when defining the headers, I made sure that the header value matches the key from the en.json file. For example, in en.json, we have "clientName", and in column.tsx, the header should be set to "clientName".
3. In the datatable, it’s simple. I use useTranslations() to get the translations and then use it in the table header. Here’s the code for that:
<TableHeader>
  {table.getHeaderGroups().map((headerGroup) => (
    <TableRow key={headerGroup.id}>
      {headerGroup.headers.map((header) => (
        <TableHead key={header.id}>
          {header.isPlaceholder
            ? null
            : flexRender(
                t(header.column.columnDef.header), // Use t here
                header.getContext()
              )}
        </TableHead>
      ))}
    </TableRow>
  ))}
</TableHeader>
I hope this helps. It’s a simple solution

**Reply by fitimbytyqi** [reply]
While this approach is really smart, you'll loose the typescript suggestion ability when typing the message.

Also this won't work when you make the header a function that returns jsx (needed when adding sorting ability or anything else).

**Reply by sidimed-bahan** [reply]
Thanks for the feedback! I understand that my solution loses TypeScript suggestions when typing the message.
As for the second point, you're right that it wouldn't work if the header needs to return JSX (for sorting or other features). In my case, I don't need that functionality right now, so this approach works for me.
Have you found a better solution ?

**Reply by fitimbytyqi** [reply]
I have 2 solutions that should work, first one is mentioned above from me

The second approach would be to utilise the meta attribute from tanstack/table.

Inside useReactTable, you pass the t function in meta attribute:

meta: {t}

in the cell from columns.tsx you can access t by using row.meta.t if I remember correctly the obj of meta.

The latter seems more stable but not sure performance wise.

**Reply by just-abdou** [reply]
I am currently working on option 1. I tried option 2 but couldn't find the row.meta.

Also, is there a way to translate some of the fields in the table rows, such as statuses?
Thanks.

**Reply by fitimbytyqi** [reply]
@just-abdou You find the meta at table.options.meta

**Reply by issam-seghir** [reply, marked as answer]
@fitimbytyqi option two really smart idea , i use it and its work perfectly 
example : 

`brands-table.tsx`

```tsx
export function BrandsTable({ columns, data }: DataTableProps) {
    const t = useTranslations();
   ...
    const table = useReactTable({
        data,
        meta: { t },
        columns,
        state: {
            sorting,
            columnVisibility,
            rowSelection,
            columnFilters,
        },
        enableRowSelection: true,
        ...
    });
....
```

`brands-columns.tsx`

```tsx
 {
        accessorKey: "username",
        header:  ({ column ,table}) => {
            return (
                <DataTableColumnHeader
                    column={column}
                    title={table.options.meta.t(
                        "BrandsPage.Table.Columns.username",
                    )}
                />
            );
        },
        cell: ({ row }) => (
            <LongText className="max-w-36">{row.getValue("username")}</LongText>
        ),
        meta: {
            className: cn(
                "drop-shadow-[0_1px_2px_rgb(0_0_0_/_0.1)] dark:drop-shadow-[0_1px_2px_rgb(255_255_255_/_0.1)] lg:drop-shadow-none ",
                "transition-colors duration-200 group-hover/row:bg-muted group-data-[state=selected]/row:bg-muted",
                "sticky start-6 z-[3] md:table-cell",
            ),
        },
        enableHiding: false,
    },
```



`en.json`

```json

 "BrandsPage": {
        "title": "Brands",
        "ActionButton": "Add Brand",
        "Table": {
            "Columns": {
                "username": "Username",
                "name": "Name",
                "email": "Email",
                "status": "Status",
                "role": "Role",
            }
        }
    },

```

**Reply by fitimbytyqi** [reply]
@issam-seghir looks great! just not sure performance wise

**Reply by Re9iNee** [reply]
in case someone needed help for type safety


```typescript, tanstack-table.d.ts
// tanstack-table.d.ts
import type { Dictionary } from "@/types/global.type";
import type { RowData } from "@tanstack/react-table";

declare module "@tanstack/react-table" {
  interface TableMeta<TData extends RowData> {
    t: Dictionary;
  }
}
```
